dnsapi/dns_freedns.sh

```diff
@@ -0,0 +1,362 @@
+#!/usr/bin/env sh
+
+#This file name is "dns_freedns.sh"
+#So, here must be a method dns_freedns_add()
+#Which will be called by acme.sh to add the txt record to your api system.
+#returns 0 means success, otherwise error.
+#
+#Author: David Kerr
+#Report Bugs here: https://github.com/Neilpang/acme.sh
+#
+########  Public functions #####################
+
+# Export FreeDNS userid and password in folowing variables...
+#  FREEDNS_USER=username
+#  FREEDNS_PASSWORD=password
+# login cookie is saved in acme account config file so userid / pw
+# need to be set only when changed.
+
+#Usage: dns_freedns_add   _acme-challenge.www.domain.com   "XKrxpRBosdIKFzxW_CT3KLZNf6q0HG9i01zxXp5CPBs"
+dns_freedns_add() {
+  fulldomain="$1"
+  txtvalue="$2"
+
+  _info "Add TXT record using FreeDNS"
+  _debug "fulldomain: $fulldomain"
+  _debug "txtvalue: $txtvalue"
+
+  if [ -z "$FREEDNS_USER" ] || [ -z "$FREEDNS_PASSWORD" ]; then
+    if [ -z "$FREEDNS_COOKIE" ]; then
+      _err "You did not specify the FreeDNS username and password yet."
+      _err "Please export as FREEDNS_USER / FREEDNS_PASSWORD and try again."
+      return 1
+    fi
+    using_cached_cookies="true"
+  else
+    FREEDNS_COOKIE="$(_freedns_login $FREEDNS_USER $FREEDNS_PASSWORD)"
+    if [ -z "$FREEDNS_COOKIE" ]; then
+      return 1
+    fi
+    using_cached_cookies="false"
+  fi
+
+  _debug "FreeDNS login cookies: $FREEDNS_COOKIE (cached = $using_cached_cookies)"
+
+  _saveaccountconf FREEDNS_COOKIE "$FREEDNS_COOKIE"
+
+  htmlpage="$(_freedns_retrieve_subdomain_page $FREEDNS_COOKIE)"
+  if [ "$?" != "0" ]; then
+    if [ "$using_cached_cookies" = "true" ]; then
+      _err "Has your FreeDNS username and password channged?  If so..."
+      _err "Please export as FREEDNS_USER / FREEDNS_PASSWORD and try again."
+    fi
+    return 1
+  fi
+
+  # split our full domain name into two parts...
+  top_domain="$(echo $fulldomain | rev | cut -d. -f -2 | rev)"
+  sub_domain="$(echo $fulldomain | rev | cut -d. -f 3- | rev)"
+
+  # Now convert the tables in the HTML to CSV.  This litte gem from
+  # http://stackoverflow.com/questions/1403087/how-can-i-convert-an-html-table-to-csv    
+  subdomain_csv="$(echo $htmlpage \
+    | grep -i -e '</\?TABLE\|</\?TD\|</\?TR\|</\?TH' \
+    | sed 's/^[\ \t]*//g' \
+    | tr -d '\n' \
+    | sed 's/<\/TR[^>]*>/\n/Ig' \
+    | sed 's/<\/\?\(TABLE\|TR\)[^>]*>//Ig' \
+    | sed 's/^<T[DH][^>]*>\|<\/\?T[DH][^>]*>$//Ig' \
+    | sed 's/<\/T[DH][^>]*><T[DH][^>]*>/,/Ig' \
+    | grep 'edit.php?' \
+    | grep $top_domain)"
+  # The above beauty ends with striping out rows that do not have an
+  # href to edit.php and do not have the top domain we are looking for.
+  # So all we should be left with is CSV of table of subdomains we are
+  # interested in.
+
+  # Now we have to read through this table and extract the data we need
+  IFS=$'\n'
```

Comment:
Done.  I rewrote the line split logic to use cut.  Not sure if the assignment of
```
nl='
'
```
is completely portable.  An alternative would be...
`nl=$'\n'`

David